File: test/Zametek.Maths.Graphs.Primitives.Tests/Entities/EdgeTests.cs
using Shouldly;
using System;
using Xunit;

namespace Zametek.Maths.Graphs.Tests
{
    public class EdgeTests
    {
        [Fact]
        public void Edge_GivenCtor_WithNullContent_ThenThrowsArgumentNullException()
        {
            Action act = () => new Edge<int, Event<int>>(null);
            act.ShouldThrow<ArgumentNullException>();
        }

        [Fact]
        public void Edge_GivenId_ThenDelegatesToContentId()
        {
            var edge = new Edge<int, Event<int>>(new Event<int>(7));

            edge.Id.ShouldBe(7);
        }

        [Fact]
        public void Edge_GivenEquals_WhenSameId_ThenEqualEvenIfContentDiffers()
        {
            var edge1 = new Edge<int, Event<int>>(new Event<int>(7, 1, 2));
            var edge2 = new Edge<int, Event<int>>(new Event<int>(7, 3, 4));

            edge1.Equals(edge2).ShouldBeTrue();
            edge1.GetHashCode().ShouldBe(edge2.GetHashCode());
        }

        [Fact]
        public void Edge_GivenEquals_WhenDifferentId_ThenNotEqual()
        {
            var edge1 = new Edge<int, Event<int>>(new Event<int>(7));
            var edge2 = new Edge<int, Event<int>>(new Event<int>(8));

            edge1.Equals(edge2).ShouldBeFalse();
        }

        [Fact]
        public void Edge_GivenEquals_WhenNull_ThenNotEqual()
        {
            var edge = new Edge<int, Event<int>>(new Event<int>(7));

            edge.Equals((Edge<int, Event<int>>)null).ShouldBeFalse();
        }

        [Fact]
        public void Edge_GivenCloneObject_ThenContentIsClonedNotShared()
        {
            var content = new Event<int>(7, 1, 2);
            content.SetAsRemovable();
            var edge = new Edge<int, Event<int>>(content);

            var clone = (Edge<int, Event<int>>)edge.CloneObject();

            clone.Id.ShouldBe(7);
            clone.Content.ShouldNotBeSameAs(content);
            clone.Content.EarliestFinishTime.ShouldBe(1);
            clone.Content.LatestFinishTime.ShouldBe(2);
            clone.Content.CanBeRemoved.ShouldBeTrue();
        }
    }
}

File: rust/primitives/tests/edge_tests.rs
//! Ports of `.../Entities/EdgeTests.cs`. The null-content and null-equality
//! guards have no Rust counterpart (the argument is non-nullable by type) and
//! are omitted.

use zametek_maths_graphs_primitives::{Edge, Event};

type Ed = Edge<i32, Event<i32>>;

#[test]
fn edge_given_id_then_delegates_to_content_id() {
    let edge = Ed::new(Event::new(7));
    assert_eq!(edge.id(), 7);
}

#[test]
fn edge_given_equals_when_same_id_then_equal_even_if_content_differs() {
    let edge1 = Ed::new(Event::with_times(7, Some(1), Some(2)));
    let edge2 = Ed::new(Event::with_times(7, Some(3), Some(4)));
    assert_eq!(edge1, edge2);
}

#[test]
fn edge_given_equals_when_different_id_then_not_equal() {
    let edge1 = Ed::new(Event::new(7));
    let edge2 = Ed::new(Event::new(8));
    assert_ne!(edge1, edge2);
}

#[test]
fn edge_given_clone_then_content_is_cloned() {
    let mut content = Event::with_times(7, Some(1), Some(2));
    content.set_as_removable();
    let edge = Ed::new(content);

    let clone = edge.clone();

    assert_eq!(clone.id(), 7);
    assert_eq!(clone.content.earliest_finish_time, Some(1));
    assert_eq!(clone.content.latest_finish_time, Some(2));
    assert!(clone.content.can_be_removed());
}
